SPEC.md

```diff
@@ -0,0 +1,84 @@
+format, which output serializers, deserializers, and a function
+dispatch interface for multiple languages.
+
+
+## WireLine specification
+
+The wire line protocol is composed of two layers. `Frames`, which
+denote an entire packet of information, and `Fields`, which correspond
+to individual data structure members.
+
+### Frames
+
+Frames in XBVC are delimited using [Consistent Overhead Byte
+Stuffing](http://www.stuartcheshire.org/papers/COBSforToN.pdf) (or
+COBS for short). COBS is a really nifty technique that allows one to
+take otherwise arbitrary binary data, and strip all Zeros from it
+(really any arbitrary value, but zero is often used). This is a very
+fast, very low overhead operation (requiring something like 2 extra
+bytes for every 255 bytes of data) and allows for packets to then be
+delimited by the stripped character. In other words, it allows one to
+have an almost readline like interface for binary packet reading.
+
+
+### Fields
+
+Fields on the other hand are represented using Extensible Bit Vectors
+(EBV). EBVs are a technique commonly used in RFID communication which
+allows for arbitrarily sized integers to be represented using the
+minimum number of bytes.
+
+This has two distinct advantages:
+
+1) For many* values it is possible to achieve a bit of compression on
+the wire. For example, the value 32 stored in a 32 bit integer is
+represented by a single byte.  The value 128 only uses two
+bytes. Obviously this goes the other way as well, if you have a value
+stored in an unsigned 8 bit container that is > 128, you actually use
+MORE bytes on the wire, however in practice this is generally not an issue.
+
+2) As each byte in an EBV reserves its most significant bit to denote
+whether or not there is additional data to be accumulated, determining
+the boundary between EBVs is as easy as checking for a zero in the
+most significant bit of each byte (I'm sensing a theme...). Once a
+zero has been found, a field can be reconstructed and one can move on
+to the next field.
+
+<!-- TODO: Add pseudo code for EBV encoding/decoding -->
+
+
+For additional information, please see the [EPC Global GEN2 standard,
+appendix
+A](https://www.gs1.org/sites/default/files/docs/epc/uhfc1g2_1_1_0-standard-20071017.pdf)
+
+
+#### Floating point numbers
+
+Say it with me: Floating point is the devil. While it would be nice to
+assume that all targets represented FP numbers using the IEEE-754
+standard, I'm not a huge fan of looking like an ass. To that point,
+the XBVC project has adopted the following standard for encoding
+floating point numbers which generalizes to all targets:
+
+Floating point numbers are transmitted as two fields: A signed 32 bit
+integer, and an unsigned 32 bit integer, in that order.
+
+The first signed number is used to represent the whole number part of
+the number in question
+
+The second unsigned number represents the fractional part of the
+number as the 9 digits after the decimal point IN REVERSE.
```

Comment:
Does this then limit the size of float between:
[-2147483648.999999999 to 2147483647.999999999] ?

If so I'd think would be really important to make that range explicit in the docs -- while there''s plenty of use cases where that is practically fine, there are a likely many where that wouldn't be an acceptable tradeoff. (I could also just be missing something).

Otherwise LGTM so far. There's some very nifty techniques in here I haven't seen before.

Reply:
Indeed it does.  It would be very easy to make the leading digit a signed 64 bit integer, which would remove this limit without causing a ton of pain on the wire.

Reply:
So we're basically picking an arbitrary fixed-point format with 64 bits in it (in the current case) and saying that's reasonable for general use. I wonder if it woul be useful/feasible let the user specify at compile time the fixed-point representation their float numbers would shipped over the wire e.g.

#define XBVC_FLOAT_WHOLE_BITS 16
#define XBVC_FLOAT_FRACTIONAL_BITS 48  

That would maybe give some control over the tradeoffs you want to make for a given application? Now I'm mostly just throwing stuff out here to make your life harder though.

Reply:
So, a couple of things:

1) It's not really arbitrary, it's a well defined word size with compiler support which allows us to avoid a bunch of manual shifting & masking.

2) This is 100% intended for talking with microcontrollers, not as a general purpose protocol generator.  If you're doing floating point math on microcontrollers that requires wire transmission of values > 2 billion, I'd argue that you're very much Doing it Wrong™

3) Very possible to avoid some tradeoff by being clever and range-checking before encoding, allowing smaller values for negative numbers (i.e. autoscaling the encoding container)

4) Might be better/easier to just send exponents? But needs benchmarking. 

I'm not a huge fan of this encoding scheme TBH, but it's easy to understand and we use FP VERY sparsely, so it's never been an issue.  As this is generally only used by us (and one other company that I'm aware of) and we're the only users of float, we have some flexibility to try other things as well.

Reply:
Cool, yeah wasn't meaning to be contentious with the arbitrary comment. I was just confirming that we're picking some range / level or precision and sticking with it for practicality. Understood on #2 -- would assume for precision processing you'd probably pick a fixed format anywho. I don't know many ADCs or sensors that have that kind of resolution so point conceded on it probably not being worth it. Was mainly just curious about the idea of having some knobs to control range vs precision -- was actually thinking more about smaller numbers than larger ones, but this seems like it's probably bike shed territory so long as the limitation is well known.

Reply:
That's a really good point on smaller numbers that I hadn't really considered....

I welcome the comments BTW :)